src/main/main.c

```diff
@@ -1215,12 +1215,13 @@ void re_thread_enter(void)
 		return;
 	}
 
-	re_lock(re);
+	/* check if already on re thread */
+	if (re_atomic_rlx(&re->thread_enter) ||
```

Comment:
I think re_atomic_rlx(&re->thread_enter)` is problematic, assume you have two different threads that try  `re_thread_enter`, bot should call `re_lock`, otherwise you have a race condition.

Reply:
In that case, ialso re_thread_check is useless and I have no way to figure out if I can safely call re_thread_enter without locking the app forever.